src/sa.py

```diff
@@ -1,15 +1,149 @@
 import argparse
+import fasta, fastq
+import re
 
+def getSuffixes(x: memoryview):
+    """
+    Gets all suffixes from a string
+    """
+    suffixes = [x[i:] for i in range(0, len(x))] 
+    return suffixes
+
+def getTrailingNumber(s):
+    m = re.search(r'\d+$', s)
+    return (s[:m.start()], int(s[m.start():]))
+
+def strcmp(s1, index, x):
+    max_i = min(len(s1), len(x)-index)
+    i = 0
+    while i < max_i and s1[i] == x[index+i]:
+        i += 1
+        if i >= max_i:
+            return 0
+    if s1[i] > x[index+i]: return 1
+    else: return -1
+
+
+def search(sa, pattern, genome):
+    low = 0
+    high = len(sa)-1
+
+    if low == high:
+        cmp = strcmp(pattern, sa[low], genome)
+        if cmp == 0:
+            yield sa[low] + 1
+            return
+
+    lowCompared = False
+    highCompared = False
+    while True:
+        mid = (high+low)//2
+        midSuffix = sa[mid]
+        cmp = strcmp(pattern, midSuffix, genome)
+
+        if cmp==0:
+            while cmp==0 and len(pattern) > len(genome) - midSuffix:
+                mid += 1
+                if mid >= len(sa): return
+                midSuffix = sa[mid]
+                cmp = strcmp(pattern, midSuffix, genome)
+            
+            oldMid = mid
```

Comment:
This `oldMid` is set to the latest `mid`, so the highest match i you do the `while` loop above it (and you don't `return` when you shouldn't). Shouldn't it be the `(high+low)//2` midpoint you computed at the beginning of the `while True` loop?